[iOS] <option> label attribute is not rendered

pxlcoder opened the pull request:
#### a232aa69a5c61fab67aa22472c48df37ba6e038b
<pre>
[iOS] &lt;option&gt; label attribute is not rendered
<a href="https://bugs.webkit.org/show_bug.cgi?id=200469">https://bugs.webkit.org/show_bug.cgi?id=200469</a>
rdar://53989128

Reviewed by Wenson Hsieh.

Use `HTMLOptionElement::displayLabel` rather than `::text` to supply the
option&apos;s title to the UI process. This method is spec-compliant, as it uses
the label attribute when available, and uses the element&apos;s text content in
quirks mode.

Add testing hooks to validate the behavior.

* LayoutTests/fast/forms/ios/select-option-label-expected.txt: Added.
* LayoutTests/fast/forms/ios/select-option-label-quirks-mode-expected.txt: Added.
* LayoutTests/fast/forms/ios/select-option-label-quirks-mode.html: Added.
* LayoutTests/fast/forms/ios/select-option-label.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.selectMenuItems):
* Source/WebCore/html/HTMLOptionElement.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _contentsOfUserInterfaceItem:]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.h:
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.mm:
(-[WKFormSelectControl menuItemTitles]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectPicker.mm:
(-[WKSelectPicker menuItemTitles]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):

Canonical link: <a href="https://commits.webkit.org/264229@main">https://commits.webkit.org/264229@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/570dab381a4ce14c9a839471a36fed5c331e253c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7533 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8723 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7334 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7119 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10231 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7236 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7887 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8828 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5246 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6461 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6559 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9438 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7045 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6370 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1676 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10604 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.h:
```diff
@@ -43,12 +43,16 @@ CGFloat adjustedFontSize(CGFloat textWidth, UIFont *, CGFloat initialFontSize, c
 - (BOOL)selectFormAccessoryHasCheckedItemAtRow:(long)rowIndex;
 @property (nonatomic, readonly) NSString *selectFormPopoverTitle;
+
+- (NSArray<NSString *> *)menuItemTitles;
```
Nit - I think we have a slight preference for:

```
@property (nonatomic, readonly) NSArray<NSString *> *menuItemTitles;
```

pxlcoder replied:
Done. I've kept the protocol method the same though, since I think it's less common to use properties there (though open to changing if desired).

pxlcoder committed:
[iOS] <option> label attribute is not rendered
https://bugs.webkit.org/show_bug.cgi?id=200469
rdar://53989128

Reviewed by Wenson Hsieh.

Use `HTMLOptionElement::displayLabel` rather than `::text` to supply the
option's title to the UI process. This method is spec-compliant, as it uses
the label attribute when available, and uses the element's text content in
quirks mode.

Add testing hooks to validate the behavior.

* LayoutTests/fast/forms/ios/select-option-label-expected.txt: Added.
* LayoutTests/fast/forms/ios/select-option-label-quirks-mode-expected.txt: Added.
* LayoutTests/fast/forms/ios/select-option-label-quirks-mode.html: Added.
* LayoutTests/fast/forms/ios/select-option-label.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.selectMenuItems):
* Source/WebCore/html/HTMLOptionElement.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _contentsOfUserInterfaceItem:]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.h:
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.mm:
(-[WKFormSelectControl menuItemTitles]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectPicker.mm:
(-[WKSelectPicker menuItemTitles]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):

Canonical link: https://commits.webkit.org/264229@main
